wfe2/wfe.go

```diff
@@ -1204,15 +1208,61 @@ func (wfe *WebFrontEndImpl) Certificate(ctx context.Context, logEvent *web.Reque
 		return
 	}
 
-	response.Header().Set("Content-Type", "application/pem-certificate-chain")
-	response.WriteHeader(http.StatusOK)
-
-	pem := pem.EncodeToMemory(&pem.Block{
+	leafPEM := pem.EncodeToMemory(&pem.Block{
 		Type:  "CERTIFICATE",
 		Bytes: cert.DER,
 	})
-	// TODO(#3291): Serve entire chain instead of just end-entity certificate.
-	if _, err = response.Write(pem); err != nil {
+
+	var responsePEM []byte
+
+	// If the WFE is configured with CertificateChains, construct a chain for this
+	// certificate using its AIA Issuer URL.
+	if len(wfe.CertificateChains) > 0 {
+		parsedCert, err := x509.ParseCertificate(cert.DER)
+		if err != nil {
+			// If we can't parse one of our own certs there's a serious problem
+			wfe.sendError(response, logEvent, probs.ServerInternal(
+				fmt.Sprintf(
+					"unable to parse Boulder issued certificate with serial %#v",
+					serial),
+			), err)
+			return
+		}
+
+		// TODO(@cpu): At present we only use the **first** AIA Issuer URL on the
```

Comment:
Let's just decide we'll only ever have one, turn this into a NOTE, and file an issue to enforce in the CA that we only ever configure one AIA Issuer URL.

Reply:
Done: https://github.com/letsencrypt/boulder/issues/3374